main/src/ca/uwaterloo/flix/language/phase/Lowering.scala

```diff
@@ -420,18 +422,43 @@ object Lowering extends Phase[Root, Root] {
     case Expression.NewChannel(exp, tpe, eff, loc) =>
       val e = visitExp(exp)
       val t = visitType(tpe)
-      Expression.NewChannel(e, t, eff, loc)
+
+      val enumSym = Symbol.mkEnumSym("Channel.ChannelImpl")
+      val chanType = Type.mkEnum(enumSym, List(t))
+      val defTpe = Type.mkImpureArrow(Type.Int32, chanType)
+      val sym = Symbol.mkDefnSym("Channel.newWithCapacity")
+      // TODO: Q: Is `loc` in the following line the location of the definition of the function or the use of the function? If it's the former, then I think that this is incorrect.
+      val callExp = Expression.Def(sym, defTpe, loc)
+      val args = List(e)
+      Expression.Apply(callExp, args, chanType, eff, loc)
 
     case Expression.GetChannel(exp, tpe, eff, loc) =>
+      // <- exp ... Here "tpe" its the element type of the channel.
+      // specifically we also know that exp itself has type Channel[tpe].
+      // we want to construct the call "Channel.get(exp)": tpe.
+      // In fact what we will care about is the type of "Channel.get"
       val e = visitExp(exp)
       val t = visitType(tpe)
-      Expression.GetChannel(e, t, eff, loc)
-
-    case Expression.PutChannel(exp1, exp2, tpe, eff, loc) =>
-      val e1 = visitExp(exp1)
-      val e2 = visitExp(exp2)
+      val enumSym = Symbol.mkEnumSym("Channel.ChannelImpl")
+      val defTpe = Type.mkImpureArrow(Type.mkEnum(enumSym, List(tpe)), tpe)
```

Comment:
Woops. This is where the bug is.

The type of `put` is `(ChannelImpl[t], t) -> ChannelImpl[t]`. But what you have here is just `ChannelImpl[t] -> ChannelImpl[t]`. You need something like:

```
val t = visitType(tpe)
val channelType = Type.mkEnum(enumSym, List(t))
Type.mkImpureUncurriedArrow(channelType, t), channelType)
```

**Note also that you should call visitType on tpe**

Reply:
Argh. I was reading the diff wrong when going over it again.